### ISSUE TITLE
csv import error setting amount-in and amount-out

### ISSUE THREAD

**Issue by mbeppler** [simonmichael/hledger#1970]
A csv file containing stock trades with number of shares, price and cost.
I would like hledger verify all the numbers.

    2022-12-12,name1,10,123.45,1234.50
    2022-12-14,name2,20,11.11,222.20

The simple rule file

    date %1

    amount-in  %3 "%2" @ %4 EUR
    amount-out  -%5 EUR

produces the following error:

    hledger: Error: in CSV rules:
    While processing CSV record: "2022-12-12","name1","10","123.45","1234.50"
    while calculating amount for posting 1
    rule "amount-in %3 "%2" @ %4 EUR" assigned value "10 "name1" @ 123.45 EUR"
    rule "amount-out -%5 EUR" assigned value "-1234.50 EUR"

    Multiple non-zero amounts were assigned for an amount field.
    Please ensure just one non-zero amount is assigned, perhaps with an if rule.
    See also: https://hledger.org/hledger.html#setting-amounts
    (hledger manual -> CSV format -> Tips -> Setting amounts)

In the manual under point 3 I find **Assign to amount (or to amount-in and amount-out).**
I interpret this as "it should work".

But a issue search found this [statement](https://github.com/simonmichael/hledger/pull/1095#issuecomment-541304477)

    1. in is NOT empty , out in NOT empty. Impossible to interpret, and we will error out

This I don't understand. If both are set hledger csv import should (**and does!** - see below) test the generated statement.

Also for statements with taxes and costs I generated entries with 3 and more accounts. And this simple test worked.

rules:

    date %1

    amount1  %3 "%2" @ %4 EUR
    amount2  0
    amount3  -%5 EUR

result:

    2022-12-12
        expenses:unknown    10 "name1" @ 123.45 EUR
        expenses:unknown                          0
        income:unknown                 -1234.50 EUR

    2022-12-14
        expenses:unknown    20 "name2" @ 11.11 EUR
        expenses:unknown                         0
        income:unknown                 -222.20 EUR

_Optimising_ the rules to

    amount1  %3 "%2" @ %4 EUR
    amount2  -%5 EUR

or

    amount1-in  %3 "%2" @ %4 EUR
    amount2-out  %5 EUR

generates the desired result:

    2022-12-12
        expenses:unknown    10 "name1" @ 123.45 EUR
        income:unknown                 -1234.50 EUR

    2022-12-14
        expenses:unknown    20 "name2" @ 11.11 EUR
        income:unknown                 -222.20 EUR

Changing the cost to 1234.56 will produce the desired error.

    hledger: Error: /Users/markus/repos/finance/test.csv:1-0:
    1 | 2022-12-12
    |     expenses:unknown    10 "name1" @ 123.45 EUR
    |     income:unknown                 -1234.56 EUR

    This multi-commodity transaction is unbalanced.
    The real postings' sum should be 0 but is: -0.06 EUR
    Consider adjusting this entry's amounts, adding missing postings,
    or recording conversion price(s) with @, @@ or equity postings.

## long story short

I'm using: hledger 1.28

Except for _amount-in_ and _amount-out_ all works as expected.

In my point of view the documentation is correct and the unexpected behaviour is due the _wrong_ first point of this [statement](https://github.com/simonmichael/hledger/pull/1095#issuecomment-541304477).

**Comment by simonmichael** [comment]
Thanks for the report. The error message is telling the truth here.. you are not allowed to have a non-zero amount for both amount-in and amount-out. (How would it choose between them ?) 

Where did you look in the manual ? Here's the first place https://hledger.org/dev/hledger.html#amount-field

The relevant details are at
https://hledger.org/dev/hledger.html#setting-amounts > c . I'll link this more clearly.

In your second issue, I'm not seeing a problem ?

**Comment by mbeppler** [comment]
Sorry that I was not able to write a more clear issue.

I want to use the _count_, _price_ and _cost_ from a csv file and **check** the values.
For this _amount-in_ and _amount-out_ needs to be set.

>    (How would it choose between them ?)

Not choose, use both and check if the balance is 0 - as is done in case of a _hand written transaction_.
This is possible with different _amount names_ (see second issue).

>     Where did you look in the manual ?

Where the error message pointed me to 😉 - https://hledger.org/dev/hledger.html#setting-amounts point 3.

There is no second _issue_. It is a _statement_ that what I want is possible if I use **amount1-in** (or amount1) and **amount2-out** (or amount2) instead of amount-in and amount-out.

This led me to the assumption that the error message is _wrong_ and not the documentation.

I hope this makes it more understandable.

**Comment by simonmichael** [comment]
Ah - you were trying to set the first posting amount with `amount-in` and the second posting amount with `amount-out`. But they are supposed to be used together for the same posting, producing only one amount from two fields. Does that explain the error ?

These rules work better:
```
date %1
amount1  %3 "%2" @ %4 EUR
amount2  -%5 EUR
```
Or, this legacy unnumbered syntax, which is equivalent to the above:
```
date %1
amount  %3 "%2" @ %4 EUR
```

**Comment by mbeppler** [comment]
Thank you for your clarification. I still don't understand the purpose or usage of _amount-in_ and _amount-out_. But I think this is due English as a third language.

As I was using _amount1_ till _amountN_ already for some transactions I used this here as a workaround. From your answer I get this is the correct way anyhow.

Thank you for your patient help and hledger of course.

**Comment by simonmichael** [comment]
Thanks @mbeppler, your feedback is useful in showing where docs can be improved. CSV rules are hard to document but I think we can do better.
https://hledger.org/dev/hledger.html#amount-field
https://hledger.org/dev/hledger.html#setting-amounts

Maybe this helps: when you have a CSV with separate columns for debits/credits, eg:
```csv
date, description, debit, credit
2023-01-01, foo, 10, 
2023-01-02, bar, , 20
```
it's impossible to parse with the normal `amount*` rules; that's when you use `amount*-in` and `amount*-out`, assigning them to the debit and credit column respectively.

**Comment by mbeppler** [comment]
After glancing at your example for some time I've got it. Tank you very much!

I import from three different banks and some other companies. All use one signed column for the _amount_.
In fact I have never seen these _two column use_. This is probably like the use of _imperial_ or _metric_ system.

And the sentence that through me of in the first place

> Assign to amount (or to amount-in **and** amount-out).

makes perfectly sense now!

Thanks again @simonmichael for not giving up on me!
